### PR TITLE
Muon Phasequad Table

### LINKED ISSUE #29611: Add PhaseQuad table
On the phase quad tab we want to add a table similar to the one used in the group/pair tab. We can then also add phasequad as a group/pair option. The phasequad is similar to a pair. 

Will need to determine the design

### PR DESCRIPTION
**Pull request by Matt-Cumber**
**Description of work.**
In the phase tab of Muon Analysis:
Removed ui file and replaced with code.
Added phasetable selector
Added table of phasequads

**To test:**
open Muon Analysis and load MUSR62260 
(it may help here to remove all the groups from the analysis in the grouping tab)
In the phase tab, try add a phasequad by pressing the plus at the bottom of the table (should not be allowed)
Now create a phasetable
Add a phasequad
Untick and tick analyse for both real and imaginary parts
Add a few more
Try pressing the - button to remove the last row, or select some rows and click the - button to remove selected rows
With some existing phasequads, create a new different phasetable and then use the phase table selector to change table (the phasequads should update)
try other stuff with groups and pairs to make sure nothing is broken there

Fixes #29611 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by AnthonyLim23** [comment]
Did you want to talk about this at some point?

**Comment by Matt-Cumber** [comment]
> Did you want to talk about this at some point?

Please could you test the changes to see if the calculations are as expected, as I am not sure if creating a new phasetable is working as expected or not. Then I would like to discuss where to go from there and what needs hooking up to https://github.com/mantidproject/mantid/blob/master/scripts/Muon/GUI/MuonAnalysis/muon_analysis_2.py#L399-L403

**Comment by AnthonyLim23** [comment]
If I load one run, create a phasequad
The select the phasequad for analyse
If I then change the runs, it does not automatically plot the selected phasequads (they are generated)

**Comment by AnthonyLim23** [comment]
I loaded run MUSR 62260
Created a phasequad 
Some of the groups disappear for a bit (they did come back but not sure how to do it)

**Comment by AnthonyLim23** [comment]
when its recalculating because we switch phase table, we should disable the GUI

**Comment by AnthonyLim23** [comment]
Create a phasequad
Change the deadtime table selection - plots vanish

**Comment by AnthonyLim23** [comment]
we still cant do multiple periods, but I think thats a different issue anyway

**Comment by AnthonyLim23** [comment]
It would be nice if we had 2 analyse boxes, one for Re and one for Im part

**Comment by AnthonyLim23** [comment]
just had a go and it looks good. The first/last good does not seem to alter the data. Is that correct (i.e. we will fix it elsewhere)?

**Comment by Matt-Cumber** [comment]
> just had a go and it looks good. The first/last good does not seem to alter the data. Is that correct (i.e. we will fix it elsewhere)?

Should there be a significant difference? The table should be recalculated and applied so the question is is the new table not being applied to the phasequads or is the difference not noticeable?

> All it needs is a call to `CropWorkspace` before the phaseqaud call

Could you show me where you mean?

**Comment by AnthonyLim23** [comment]
here we apply the deadtime before to the input of the phasequad https://github.com/mantidproject/mantid/blob/17a647c1998d22fb926afbfabdf2475162ab6625/scripts/Muon/GUI/Common/contexts/muon_context.py#L306

we just want to apply a crop between the deadtime and the phasequad. So the input has a smaller range (ignores data we dont like).

**Comment by AnthonyLim23** [comment]
Crop works as expected. We do need to use the time zero, but that will need us to talk to James for details on how to use it. So lets make an issue for it.

**Comment by AnthonyLim23** [comment]
Opening the frequency domain GUI causes a crash

File: scripts/Muon/GUI/Common/phase_table_widget/phase_table_view.py [hunk 1]
```diff
@@ -196,23 +300,108 @@ def setup_phase_table_options_table(self):
...
+        self.phase_table_selector_combo.blockSignals(True)
+        index = self.phase_table_selector_combo.findText(table)
+        if index != -1:
+            self.phase_table_selector_combo.setCurrentIndex(index)
+        self.phase_table_selector_combo.blockSignals(False)
+
+    def get_table_contents(self):
+        if self._updating:
+            return []
+        ret = [None for _ in range(self.num_rows())]
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/phase_table_widget/phase_table_view.py:397]: whats ret?

**Comment by AnthonyLim23** [comment]
It all seems to work as expected. Except for one bug. When I create a phasequad in FDA, it is plotted on the fit pane (should only show frequency domain data) and the MaxEnt I calculated, does not show up.

**Comment by AnthonyLim23** [comment]
can you rebase from master? The plotting breaks because of the plotting changes.

File: scripts/Muon/GUI/MuonAnalysis/muon_analysis_2.py [hunk 2]
```diff
@@ -415,6 +412,9 @@ def setup_on_recalulation_finished_notifier(self):
         for observer in self.plot_widget.data_changed_observers:
             self.grouping_tab_widget.group_tab_presenter.calculation_finished_notifier.add_subscriber(observer)
 
+        self.phase_tab.phase_table_presenter.calculation_finished_notifier.add_subscriber(
+            self.update_plot_observer)
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/MuonAnalysis/muon_analysis_2.py:416]: this line causes a crash, same in FDA

**Review by AnthonyLim23 — DISMISSED** [review]
looks good to me and it all works as expected :shipit:

**Comment by Matt-Cumber** [comment]
@AnthonyLim23 I have rebase on master to resolve a merge conflict

**Review by AnthonyLim23 — APPROVED** [review]
Looks good. :shipit:

**Review by DavidFair — APPROVED** [review]
LGTM,

One piece of feedback for the future, I'd **strongly** recommend using a context manager in Python for things like 

```python
self.view.disable_update()
...
self.view.enable_updates()
```

If anything throws, or returns unexpectedly...etc. your GUI will be in a bugged state. Instead adding a context manager takes all the risk away:

```python
@contextmanager
def disable_updates(presenter)
    presenter.view.disable_updates()
    try:
        yield
    finally:
        presenter.view.enable_updates()

class PhaseTablePresenter:
    def someMethod(self):
        with disable_updates(self):
            some_call_that_needs_no_updates()
            another_call()
```

`self.view.enable_updates()` will always be called unless Python dies a horrible death, in which case you've got bigger problems anyway

**Pull request merged**